### PR TITLE
Fix AWSKeyValueStore for crashes in Encryption Key management

### PR DESCRIPTION
**Pull request by kvasukib**
*Issue #, if available:*

#937 
#1115 
#1109 
#1108 

*Description of changes:*

**Causes of Failures**

1. The retrieval of secrets from disk fails due to the encryption key being lost. 

     a. End-user changes lock screen type which wipes off the encryption keys in the Android KeyStore. This could happen when the app is in foreground, background, closed.

2. The retrieval of secrets from disk fails due to the wrong encryption key used.

    a. End-user uninstalls the app which wipes off the encryption key in Android KeyStore. The encrypted secrets are backed up by Android and restored when the app is installed again. No

3. The retrieval of secrets from disk fails due to the encryption key not being recoverable.

    a. A cause has not been identified why key is not recoverable.

**Proposed Fix**

Under all these circumstances, the `AWSKeyValueStore` would return `null` for `get` operations and not write to disk for `put` operations. All the consumers of the `AWSKeyValueStore` will be updated with improved error handling logic and retrieve new set of credentials based on the implementation.

1. Amazon Cognito Federated Identities - COMPLETED
2. Amazon Cognito User Pools - COMPLETED 
3. Amazon Cognito Auth - COMPLETED
4. AWSMobileClient - COMPLETED

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserSession.java [hunk 1]
```diff
@@ -103,16 +118,36 @@ public boolean isValid() {
     }
 
     /**
-     * Returns true if this session for the threshold set in {@link CognitoIdentityProviderClientConfig#refreshThreshold}.
+     * Returns true if this session for the threshold set in
```
Comment (palpatim) [inline aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserSession.java]: "if this session is valid for the threshold..."

File: aws-android-sdk-core-test/src/androidTest/java/com/amazonaws/internal/keyvaluestore/AWSKeyValeStoreNoCachingIntegrationTest.java [hunk 2]
```diff
@@ -38,9 +38,9 @@
 import java.util.ArrayList;
 import java.util.List;
 
-public class AWSKeyStoreNoCachingIntegrationTest  extends CoreIntegrationTestBase {
+public class AWSKeyValeStoreNoCachingIntegrationTest extends CoreIntegrationTestBase {
```
Comment (palpatim) [inline aws-android-sdk-core-test/src/androidTest/java/com/amazonaws/internal/keyvaluestore/AWSKeyValeStoreNoCachingIntegrationTest.java]: s/AWSKeyValeStoreNoCachingIntegrationTest/AWSKeyVal`u`eStoreNoCachingIntegrationTest/

File: aws-android-sdk-core/src/main/java/com/amazonaws/auth/CognitoCachingCredentialsProvider.java [hunk 3]
```diff
@@ -95,7 +95,7 @@
 
     private static final String TAG = "CognitoCachingCredentialsProvider";
 
-    private AWSKeyValueStore awsKeyValueStore;
+    AWSKeyValueStore awsKeyValueStore;
```
Comment (palpatim) [inline aws-android-sdk-core/src/main/java/com/amazonaws/auth/CognitoCachingCredentialsProvider.java]: Why the change?

File: aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/AWSMobileClient.java [hunk 4]
```diff
@@ -258,8 +258,8 @@
     AWSMobileClientCognitoIdentityProvider provider;
     DeviceOperations mDeviceOperations;
     AmazonCognitoIdentityProvider userpoolLL;
-    private Auth hostedUIJSONConfigured;
-    private Auth hostedUI;
+    Auth hostedUIJSONConfigured;
```
Comment (palpatim) [inline aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/AWSMobileClient.java]: were these the changes you were going to revert?

File: aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserSession.java (hunk 1 above)
Comment (kvasukib) [reply aws-android-sdk-cognitoidentityprovider/src/main/java/com/amazonaws/mobileconnectors/cognitoidentityprovider/CognitoUserSession.java]: Fixed

File: aws-android-sdk-core-test/src/androidTest/java/com/amazonaws/internal/keyvaluestore/AWSKeyValeStoreNoCachingIntegrationTest.java (hunk 2 above)
Comment (kvasukib) [reply aws-android-sdk-core-test/src/androidTest/java/com/amazonaws/internal/keyvaluestore/AWSKeyValeStoreNoCachingIntegrationTest.java]: Fixed. Thanks

File: aws-android-sdk-core/src/main/java/com/amazonaws/auth/CognitoCachingCredentialsProvider.java (hunk 3 above)
Comment (kvasukib) [reply aws-android-sdk-core/src/main/java/com/amazonaws/auth/CognitoCachingCredentialsProvider.java]: I am using this for testing. So made it package private.

File: aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/AWSMobileClient.java (hunk 4 above)
Comment (kvasukib) [reply aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/AWSMobileClient.java]: Reverted. Sorry it was a miss.
